src/rosdiscover/observer/observer.py

```diff
@@ -0,0 +1,45 @@
+# -*- coding: utf-8 -*-
+__all__ = ("Observer",)
+import contextlib
+from abc import ABC, abstractmethod
+from typing import Dict, Iterator
+import roswire
+from roswire import AppInstance, ROSVersion
+
+from ..config import Config
+from ..interpreter import NodeContext, SystemSummary
+
+
+class Observer(ABC):
+
+    @classmethod
+    @contextlib.contextmanager
+    def for_container(cls,
+                      container: str,
+                      config: Config,
+                      ) -> Iterator['Observer']:
+        """Constructs and interpreter for a given running container"""
+        rsw = roswire.ROSWire()
+        app = rsw.app(config.image, config.sources)
+        instance = app.attach(container, require_description=True)
+        if app.description.distribution.ros == ROSVersion.ROS1:
+            from .ros1 import ROS1Observer
+            yield ROS1Observer(instance, config)
+        else:
+            from .ros2 import ROS2Observer
+            yield ROS2Observer(instance, config)
+
+    def __init__(self, app: AppInstance, config: Config):
+        self._app_instance = app
+        self._config = config
+        self._nodes: Dict[str, NodeContext] = {}
```

Comment:
This isn't really necessary or used.

Reply:
`self._nodes` isn't necessary here. We should try to ensure that `Observer` is a stateless class (since it needn't maintain state to implement its API). Maintaining internal state adds unnecessary complexity to the implementation and makes the code harder to test, debug, and maintain (method calls are no longer idempotent).

Reply:
Also, don't forget to add the return type annotation (`-> None`).

Reply:
Done.